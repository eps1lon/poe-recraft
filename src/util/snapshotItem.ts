import {
  Popup as ItemPopup,
  Rarity as ComponentRarity
} from 'poe-components-item';
import { Item } from 'poe-mods';

// returns an object with all props for poe-components-item/popup
export default function snapshotItem(item: Item): ItemPopup['props']['item'] {
  return {
    base: {
      name: item.baseitem.name
    },
    rarity: mapRarity(item)
  };
}

function mapRarity(item: Item): ComponentRarity {
  if (item.rarity.isNormal()) {
    return ComponentRarity.normal;
  } else if (item.rarity.isMagic()) {
    return ComponentRarity.magic;
  } else if (item.rarity.isRare()) {
    return ComponentRarity.rare;
  } else if (item.rarity.isUnique()) {
    return ComponentRarity.unique;
  } else {
    throw new Error('unrecognized item rarity');
  }
}
