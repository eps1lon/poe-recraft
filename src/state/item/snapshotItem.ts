import { Popup as ItemPopup, Rarity } from 'poe-components-item';
import { formatStats } from 'poe-i18n';
import { Container, Item, ArmourProperties } from 'poe-mods';

import { namei18n } from '../../util/item';

/**
 * creates a "dumb" object with view relevant properties from item
 *
 * can be used for poe-components-item#ItemPopup
 * @param item
 */
export default function snapshotItem(
  item: Item,
  descriptions: {},
  messages: {}
): ItemPopup['props']['item'] {
  let rarity: Rarity;
  if (item.rarity.isNormal()) {
    rarity = Rarity.normal;
  } else if (item.rarity.isNormal()) {
    rarity = Rarity.normal;
  } else if (item.rarity.isMagic()) {
    rarity = Rarity.magic;
  } else if (item.rarity.isRare()) {
    rarity = Rarity.rare;
  } else if (item.rarity.isUnique()) {
    rarity = Rarity.unique;
  } else {
    throw new Error('unrecognized rarity');
  }

  const properties = snapshotProperties(item);
  const [typeLineOrName, typeLine] = item.name.lines();

  return {
    base: {
      name: namei18n(item, messages)
    },
    name: typeLine !== undefined ? typeLineOrName : undefined,
    elder: item.isElderItem(),
    shaper: item.isSHaperItem(),
    implicitStats: statsTranslated(item.implicits, descriptions),
    explicitStats: statsTranslated(item.affixes, descriptions),
    rarity,
    ...properties,
    requirements: snapshotRequirements(item)
  };
}

function snapshotProperties(item: Item) {
  const { properties } = item;
  const base_properties = {
    quality: properties.quality
  };

  if (properties instanceof ArmourProperties) {
    return {
      ...base_properties,
      ...properties.defences()
    };
  } else {
    return base_properties;
  }
}

function snapshotRequirements(item: Item) {
  const { level, dex, int, str } = item.requirements.list();

  return {
    level: {
      augmented: false,
      value: level
    },
    dexterity: {
      augmented: false,
      value: dex
    },
    intelligence: {
      augmented: false,
      value: int
    },
    strength: {
      augmented: false,
      value: str
    }
  };
}

function statsTranslated(container: Container<any>, descriptions: {}) {
  return formatStats(
    Object.entries(container.stats()).map(([id, stat]) => {
      return {
        id,
        value: stat.values.valueOf()
      };
    }),
    { datas: descriptions }
  );
}
