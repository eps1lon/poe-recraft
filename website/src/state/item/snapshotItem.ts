import { PopupIntl as ItemPopup, Rarity } from 'poe-components-item';
import { Fallback, formatStats } from 'poe-i18n';
import { ArmourProperties, Container, Item, WeaponProperties } from 'poe-mods';

import { propOrUndefined } from '../../util/typesafe';

/**
 * creates a "dumb" object with view relevant properties from item
 *
 * can be used for poe-components-item#ItemPopup
 * @param item
 */
export default function snapshotItem(
  item: Item,
  descriptions: {},
): PropsType<typeof ItemPopup>['item'] {
  let rarity: Rarity;
  if (item.rarity.isNormal()) {
    rarity = 'normal';
  } else if (item.rarity.isMagic()) {
    rarity = 'magic';
  } else if (item.rarity.isRare()) {
    rarity = 'rare';
  } else if (item.rarity.isUnique()) {
    rarity = 'unique';
  } else {
    throw new Error('unrecognized rarity');
  }

  const properties = snapshotProperties(item);
  const [typeLineOrName, typeLine] = item.name.lines();
  // for magic items there can only be at most one prefix/suffix
  const prefix = propOrUndefined(
    item.affixes.getPrefixes().map(({ props }) => props)[0],
    'id',
  );
  const suffix = propOrUndefined(
    item.affixes.getSuffixes().map(({ props }) => props)[0],
    'id',
  );

  return {
    base: {
      id: item.baseitem.id,
    },
    name: typeLine !== undefined ? typeLineOrName : undefined,
    elder: item.isElderItem(),
    shaper: item.isSHaperItem(),
    implicitStats: statsTranslated(item.implicits, descriptions),
    explicitStats: statsTranslated(item.affixes, descriptions),
    rarity,
    ...properties,
    requirements: snapshotRequirements(item),
    corrupted: item.props.corrupted,
    prefix,
    suffix,
  };
}

function snapshotProperties(item: Item) {
  const { properties } = item;
  const base_properties = {
    quality: properties.quality,
  };

  if (properties instanceof ArmourProperties) {
    return {
      ...base_properties,
      ...properties.defences(),
    };
  } else if (properties instanceof WeaponProperties) {
    return {
      ...base_properties,
      ...snapshotWeaponProperties(properties),
    };
  } else {
    return base_properties;
  }
}

function snapshotWeaponProperties(properties: WeaponProperties) {
  const physical_damage = properties.physical_damage();
  const cold_damage = properties.cold_damage();
  const fire_damage = properties.fire_damage();
  const lightning_damage = properties.lightning_damage();
  const chaos_damage = properties.chaos_damage();

  return {
    physical_damage: {
      augmented: physical_damage.min.augmented || physical_damage.max.augmented,
      value: unsafeAsTuple([
        physical_damage.min.value,
        physical_damage.max.value,
      ]),
    },
    cold_damage: unsafeAsTuple([cold_damage.min.value, cold_damage.max.value]),
    fire_damage: unsafeAsTuple([fire_damage.min.value, fire_damage.max.value]),
    lightning_damage: unsafeAsTuple([
      lightning_damage.min.value,
      lightning_damage.max.value,
    ]),
    chaos_damage: unsafeAsTuple([
      chaos_damage.min.value,
      chaos_damage.max.value,
    ]),
    aps: properties.attack_speed(),
    crit: properties.crit(),
    range: properties.weapon_range(),
  };
}

function snapshotRequirements(item: Item) {
  const { level, dex, int, str } = item.requirements.list();

  return {
    level,
    dexterity: dex,
    intelligence: int,
    strength: str,
  };
}

function statsTranslated(container: Container<any>, descriptions: {}) {
  return formatStats(
    Object.entries(container.stats()).map(([id, stat]) => {
      return {
        id,
        value: stat.values.valueOf(),
      };
    }),
    { datas: descriptions, fallback: Fallback.id },
  );
}

/**
 * identity function
 * typecasts an array to a tuple
 * no checks at runtime so be sure to only pass explicitly i.e.
 * unsafeAsTuple([1, 2])
 * @param list
 */
function unsafeAsTuple<T>(list: T[]): [T, T] {
  return list as [T, T];
}
