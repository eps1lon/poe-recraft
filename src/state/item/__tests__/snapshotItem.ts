import { createItems } from 'poe-mods';
import snapshotItem from '../snapshotItem';
const item_props: any = 'poe-mods/data/items';

const messages = {
  baseitemtypes: require('poe-i18n/locale-data/en/BaseItemTypes.json')
};
const descriptions = {
  stat_descriptions: require('poe-i18n/locale-data/en/stat_descriptions.json')
};
const items = createItems(require('poe-mods/data/items.json'));

it('matches weapons', () => {
  const weapon = items.fromName('Platinum Kris');
  expect(snapshotItem(weapon, descriptions, messages)).toMatchSnapshot(
    'weapon'
  );
});
