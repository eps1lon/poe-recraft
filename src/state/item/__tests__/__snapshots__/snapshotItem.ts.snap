// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches weapons: weapon 1`] = `
Object {
  "base": Object {
    "name": "Metadata/Items/Weapons/OneHandWeapons/Daggers/Dagger20",
  },
  "elder": false,
  "explicitStats": Array [],
  "implicitStats": Array [
    "50% increased Global Critical Strike Chance",
  ],
  "name": undefined,
  "quality": 0,
  "rarity": 0,
  "requirements": Object {
    "dexterity": Value {
      "augmented": false,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 76,
        "min": 76,
      },
    },
    "intelligence": Value {
      "augmented": false,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 149,
        "min": 149,
      },
    },
    "level": Value {
      "augmented": false,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 64,
        "min": 64,
      },
    },
    "strength": Value {
      "augmented": false,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 0,
        "min": 0,
      },
    },
  },
  "shaper": false,
}
`;
